connectors/sync_job_runner.py

```diff
@@ -85,13 +82,16 @@ def __init__(
     async def execute(self):
         if self.running:
             raise SyncJobRunningError(f"Sync job {self.job_id} is already running.")
-
         self.running = True
-        if not await self._claim_job():
-            logger.error(
-                f"Unable to claim job {self.job_id} for connector {self.connector_id}"
+
+        if not await self.sync_starts():
```

Comment:
What does `sync_starts` mean? I read the code and can't understand it - method name is confusing.

Also the flow of the function that returns True/False confuses me too, maybe better build it on exceptions or inline the function?

Reply:
>What does sync_starts mean?

It's from the `connector.sync_starts` method. In `Connector` class, there is a `sync_starts` method (update `last_sync_status` to `in_progress`) and a `sync_done` method (update `last_sync_status` to one of the terminal statuses).

>Also the flow of the function that returns True/False confuses me too, maybe better build it on exceptions or inline the function?

The returned value indicates whether it starts the sync successfully. If we make it raise an exception, do you think we should 1) catch it in `execute` or 2) propagate it and exit `execute`?

If 1), I don't feel it's necessary to raise can catch it here. If 2), do you think this is an exception? I feel this is one of the expected results. Especially for this scenario: When there is a running job and a pending job for a connector, in each iteration, the pending job will be picked, but returned here immediately because `if self.connector.last_sync_status == JobStatus.IN_PROGRESS` is evaluated to `true`.

Reply:
Maybe it's worth to rename it to `start_sync` and `mark_as_done`?

As to catching - depends on which class needs to handle it. It can be an exception with text "Failed to start sync for connector {self.connector_id}, the sync is executed by another connector service." that is caught by the service executor and I think it'll be fine.

To me exceptions are the way to specify exceptional scenarios and this is one of them - you attempt to start a sync but fail cause someone else did. It's more of a way to express code/scenarios, I feel that method that returns whether it did something in expected manner or not with True/False is not expressive enough, but it might be purely taste thing to me

Reply:
I think it's a great idea, thanks for the elaborate explanation too!

I'll approve this PR and it can be implemented either in the scope of this one or in the scope of a different PR

Reply:
Hi @artem-shelkovnikov , I will merge the PR first, and make the change in a subsequent PR.

Reply:
PR created: https://github.com/elastic/connectors-python/pull/788